rsconnect/actions.py

```diff
@@ -1014,27 +1014,31 @@ def gather_basic_deployment_info_for_notebook(connect_server, app_store, file_na
     if new and app_id:
         raise api.RSConnectException("Specify either a new deploy or an app ID but not both.")
 
-    if app_id is not None:
+    app_mode = AppModes.STATIC
+    existing_app_mode = None
+    if app_id is None:
+        # Possible redeployment - check for saved metadata.
+        # Use the saved app information unless overridden by the user.
+        app_id, existing_app_mode = app_store.resolve(connect_server.url, app_id, app_mode)
+        logger.debug("Using app mode from app %s: %s" % (app_id, app_mode))
+    else:
         # Don't read app metadata if app-id is specified. Instead, we need
         # to get this from Connect.
         app = api.get_app_info(connect_server, app_id)
-        app_mode = AppModes.get_by_ordinal(app.get("app_mode", 0), True)
-
-        logger.debug("Using app mode from app %s: %s" % (app_id, app_mode))
-    elif static:
+        existing_app_mode = AppModes.get_by_ordinal(app.get("app_mode", 0), True)
+    if not new and existing_app_mode and app_mode != existing_app_mode:
+        msg = (
+            "Deploying with mode '%s',\n"
+            + "but the existing deployment has mode '%s'.\n"
+            + "Use the --new option to create a new deployment of the desired type."
+        ) % (app_mode.desc(), existing_app_mode.desc())
+        raise api.RSConnectException(msg)
+
+    if static:
```

Comment:
It seems like this needs to go before the conflict check above, so we're checking correct value for `app_mode`.